torch/csrc/cuda/AutoGPU.h

```diff
@@ -2,9 +2,14 @@
 #define THCP_AUTOGPU_INC
 
 #include <Python.h>
+#include "THP.h"
 #include "torch/csrc/utils/auto_gpu.h"
 
+#if defined(WITH_CUDA) && defined(_MSC_VER)
+class THP_CLASS THCPAutoGPU : public AutoGPU {
```

Comment:
Why is this `ifdef` needed? We could just always define `THP_CLASS` to no-op on non-Windows platforms

Reply:
I don't know but there's a weird linking error if THP_CLASS is used when building non CUDA versions in windows.

Reply:
And it's only in this single class? That's really weird I'd investigate that

Reply:
The reason is that `AutoGPU.cpp` is not compiled, so linking will fail in MSVC. I've fixed that by compiling `AutoGPU.cpp` and excluding those THC functions in non CUDA builds.